Complete Java lambda support [TG-11760]

smowton opened the pull request:
This adds the various missing pieces of Java lambda support:

* Allow direct method references (javac uses these when no adaptation is necessary, there are no captured parameters to pass on, etc)
* Support lambdas that use invokevirtual / invokeinterface (these are generated similar to above but for instance and interface methods, e.g. `this::f` or `someInterface::method`)
* Support the `newinvokespecial` lambda binding, used for `SomeType::new`, which means we must both instantiate and call the constructor
* Support functional interfaces whose defined method is inherited
* Support functional interfaces that also define `equals` or `hashCode`
* Support functional interfaces featuring default methods

smowton committed:
Store lambda methods kinds in java_class_typet

This paves the way for supporting multiple lambda method kinds, e.g. newinvokespecial lambdas
which instantiate and construct a type.

smowton committed:
Java lambdas: support all functional interfaces

Previously we failed to detect a functional interface when the abstract method
being implemented was inherited, or the interface also declared methods which are
defined on java.lang.Object such as toString.

smowton committed:
Java lambdas: accept direct method references

Usually an intermediate synthetic method is created by javac, but sometimes it can
use a direct reference to a method, which we would previously have failed to spot.

For example, Callable c = x -> x + 1 will generate a synthetic private method housing
the body of the lambda, but Callable c = SomeClass::someCallable may not generate one
but rather directly use SomeClass.someCallable as the lambda target. Whether this happens
in practice depends on the compiler.

smowton committed:
Java lambdas: support newinvokespecial lambdas

These encode a lambda that instantiates and constructs an instance of some type without
using an intermediate method.

smowton committed:
Activate existing lambda tests that now pass

smowton committed:
Add tests for the newly supported kinds of lambda

smowton committed:
Add test exercising static, virtual and interface lambdas

These compile down to direct references to a static, virtual and interface method (i.e. to
method reference kinds REF_invokeStatic, REF_invokeVirtual and REF_invokeInterface as described
at https://docs.oracle.com/javase/9/docs/api/java/lang/invoke/MethodHandleInfo.html)

smowton committed:
Add support for invoke-virtual and invoke-interface lambdas

To support these we store lambda descriptors in the same form as virtual method references,
using class_method_descriptor_exprt, lowering it to a symbol_exprt if we have a known callee.

This is also the first form of synthetic method to produce a virtual method call, hence the
alteration of notify_static_method_calls to tolerate them.

smowton committed:
Add method documentation for lambda body synthesis

peterschrammel commented on jbmc/src/java_bytecode/java_bytecode_convert_class.cpp:
```diff
@@ -426,7 +426,8 @@ void java_bytecode_convert_classt::convert(
       ? class_type.add_unknown_lambda_method_handle()
       : class_type.add_lambda_method_handle(
-          "java::" + id2string(lambda_entry.second.lambda_method_ref));
+          "java::" + id2string(lambda_entry.second.lambda_method_ref),
```
⛏️ `ID_java + "::"` ?

thk123 replied:
not sure - "java::" is quite a common string - probably makes more sense to pull out a ID_java_prefix than hide a few of them with this

thk123 replied:
Also unchanged by this PR

peterschrammel commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -92,6 +92,83 @@ lambda_method_handle(
+  for(const auto &method : methods)
+  {
+    const auto &name = method.get_name();
+    static irep_idt equals = "equals";
```
`const` ?

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -330,6 +329,11 @@ class method_handle_infot : public structured_pool_entryt
     reference_index = entry.ref2;
   }
 
+  method_handle_kindt get_reference_kind() const
```
:grey_question: `get_method_handle_kind`? Not sure why the change in nomenclature?

thk123 commented on jbmc/src/java_bytecode/java_types.h:
```diff
@@ -450,34 +450,62 @@ class java_class_typet:public class_typet
-  //   }
-  // };
-  using java_lambda_method_handlest = irept::subt;
+  enum class method_handle_typet
```
:bulb: `method_handle_kindt`?

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -92,6 +92,83 @@ lambda_method_handle(
+    }
+    if(!ns.lookup(name).type.get_bool(ID_C_abstract))
+      continue;
+    if(result.has_value() && name != *result)
```
Is it allowed for a functional interface to have two unimplemented methods with the same name?

smowton replied:
These are mangled names (since we expect to find them in `ns`); renamed to clarify.

thk123 replied:
Sure - but I don't understand why an interface would be allowed to have two methods with the same mangled name? I mean - surely that is just impossible? I.e. in what circumstances is `result.has_value() && name == *result`?

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -92,6 +92,83 @@ lambda_method_handle(
+  const irep_idt &interface_id,
+  const namespacet &ns)
+{
+  static irep_idt jlo = "java::java.lang.Object";
```
const here too - leaving mutable global state lying around is a recipe for bugs

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1911,22 +1911,12 @@ java_bytecode_parsert::parse_method_handle(const method_handle_infot &entry)
 
   lambda_method_handlet lambda_method_handle;
 
-  if(has_prefix(name_and_type.get_name(pool_entry_lambda), "lambda$"))
```
Could commit message include an example of when this happens:

> but sometimes it can use a direct reference to a method, which we would previously have failed to spot.

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1905,6 +1905,20 @@ java_bytecode_parsert::parse_method_handle(const method_handle_infot &entry)
+    entry.get_reference_kind() ==
+      method_handle_infot::method_handle_kindt::REF_putStatic)
+  {
+    // Field accessors not implemented yet
```
:question: I don't understand is this if you assign a lambda from a field containing a lambda?

smowton replied:
Expanded comment; in brief these are notional "field bindings", e.g. you would be able to say `Producer<Integer> = this::someIntegerField` and get a REF_getField instead of a REF_invokeSpecial pointing at a short function that fetches the field. While it made it into the JVMS and might be used in future, AFAICT nobody has actually done that.

thk123 replied:
lol cool feature I guess :sunglasses:

thk123 commented on jbmc/regression/jbmc/lambda-interface-declaring-equals/InterfaceDeclaringEquals.java:
```diff
@@ -0,0 +1,6 @@
+
+  public int f(int x);
+
+  public boolean equals(Object other);
```
Could we have a test for a functional interface that also declares hashcode

thk123 commented on jbmc/regression/jbmc/lambda-reference-kinds/test.desc:
```diff
@@ -0,0 +1,9 @@
+line 20 assertion at file Test\.java line 20 function java::Test\.main:\(\)V bytecode-index 39: SUCCESS
+line 21 assertion at file Test\.java line 21 function java::Test\.main:\(\)V bytecode-index 50: SUCCESS
+^EXIT=0$
+^SIGNAL=0$
```
:no_entry_sign: Please add an explanation as to the purpose of this test (I believe it is checking when the method reference is directly inlined?

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1881,6 +1881,12 @@ void java_bytecode_parsert::parse_local_variable_type_table(methodt &method)
+/// of handling the method requires. invokestatic/special translate into direct
+/// method calls; invokevirtual/interface translate into virtual dispatch,
+/// newinvokespecial translates into a special instantiate-and-construct
+/// sequence. The field-manipulation reference kinds appear never to happen in
+/// reality and don't have syntax in the Java language.
```
I think this info would be good to put nearer the block that handles this

smowton replied:
It's now mentioned in both places

thk123 commented on jbmc/src/java_bytecode/lambda_synthesis.cpp:
```diff
@@ -102,6 +102,10 @@ class no_unique_unimplemented_method_exceptiont : public std::exception
 };
 
+/// Find the unique abstract method that isn't equals(...) or hashCode() in a
+/// kist of methods. Returns the method's symbol-table identifier if one is
```
```suggestion
/// list of methods. Returns the method's symbol-table identifier if one is
```

thk123 reviewed (commented):
In a couple of places you allow the name of the method to match the name of the method found in the functional interface, but afaict you don't have any tests for this scenario, can they be added. 

Couple of minor points but looks good

smowton committed:
Name method handles uniformly

They are method handles not references, and their varieties are kinds not types

thk123 commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1896,33 +1926,45 @@ java_bytecode_parsert::parse_method_handle(const method_handle_infot &entry)
-    lambda_method_handle.handle_type =
-      method_handle_typet::LAMBDA_METHOD_HANDLE;
-
-    return lambda_method_handle;
```
:bulb: Given that this could happen - it might make sense to output a warning, or even an invariant

thk123 reviewed (approved):
Still a little confused about the check for the mangled name - happy to chat in slack if easier to explain

thk123 reviewed (dismissed):
Oops approved too early - could you have a look at the gaps in testing highlighted by code cov, in particular:

 - :no_entry_sign: a test for with an interface that contains two methods
 - :no_entry_sign: a test for an interface with one method, and inherits another
 - :question: a test for a constructor? - I thought I'd seen this, but there is a whole untested function: `instantiate_new_object`
 - :bulb: and ideally something to cover this line: https://codecov.io/gh/diffblue/cbmc/compare/c6082355bde3657bdfdefde8eff16a2813684b54...4b58a793061915277cf8255c628f94bff27414af/diff#D32-1244

smowton committed:
Fix comment in notify_static_method_calls

smowton committed:
Disable notify_static_method_calls path for side_effect_expr_function_callt

This is currently dead code, so replace it with an invariant describing the situation.

smowton committed:
java_class_typet: record method descriptor

Along with the basename already stored on class_typet::methodt, this means we can compare
two Java methods for name and descriptor compatibility without having to parse the composite
method identifier.

smowton committed:
Java frontend: support lambdas whose functional interfaces merge several methods

For example, if Intf1 declares int f(int) and so does Intf2, then Merge which implements both
Intf1 and Intf2 is a functional interface.

smowton committed:
Java frontend: support functional interfaces created by defining or explicitly abstracting methods

It's possible to make a functional interface by taking a parent or parents that have more than one
abstract method, then inheriting default definitions or making definitions yourself so that only one
abstract method remains, and so the interface becomes functional. Similarly you can re-declare a method
as abstract even if a parent made a default definition.

This commit adds support for all of these cases and tests them.

romainbrenguier commented on jbmc/src/java_bytecode/java_bytecode_language.cpp:
```diff
@@ -1241,7 +1241,16 @@ static void notify_static_method_calls(
+            "that has changed, remove this invariant. Also note that "
+            "as of the time of writing remove_virtual_functions did "
+            "not support this form of function call.");
+          // needed_lazy_methods->add_needed_method(fn_sym->get_identifier());
```
:pick: remove or comment with `#if 0`
